src/main/java/com/r307/arbitrader/service/paper/PaperTradeService.java

```diff
@@ -279,7 +281,25 @@ private BigDecimal getBaseDelta(LimitOrder order) {
      */
     private BigDecimal getCounterFee(LimitOrder order) {
         BigDecimal cumulativeCounterAmount = order.getAveragePrice() != null ? order.getOriginalAmount().multiply(order.getAveragePrice()) : order.getOriginalAmount().multiply(order.getLimitPrice());
-        BigDecimal feePercentage = exchangeService.getExchangeFee(exchange, order.getCurrencyPair(), false);
+        BigDecimal feePercentage = getFee(order);
         return exchangeService.getExchangeMetadata(exchange).getFeeComputation() == SERVER ? cumulativeCounterAmount.multiply(feePercentage) : BigDecimal.ZERO;
     }
+
+    private BigDecimal getFee(LimitOrder order) {
+        final String exchangeName = exchange.getExchangeSpecification().getExchangeName();
+        final ExchangeFee exchangeFee = exchangeService.getExchangeFee(exchange, (CurrencyPair) order.getInstrument(), false);
+        final ExchangeConfiguration exchangeConfiguration = exchangeService.getExchangeMetadata(exchange);
+
+        final BigDecimal fee;
+        if (order.getLeverage() == null) {
+            fee = exchangeFee.getTradeFee();
+            LOGGER.info("exchange:{}|exchange:{}|long order using {} fee", exchangeName, exchangeConfiguration.getMargin(), fee);
+        }
+        else {
+            fee = exchangeFee.getMarginFee().get();
+            LOGGER.info("exchange:{}|exchange:{}|short order using {} fee", exchangeName, exchangeConfiguration.getMargin(), fee);
+        }
```

Comment:
A few things here:
- about the logs, the paper exchange is just an exchange, it does not need to know if it's a long or short exchange, only if it's a BID or ASK order. Also the logs say `exchange:{}|exchange:{}` instead of `exchange:{}|margin:{}` I also try to start the paper trading logs by `{} paper exchange:` so we can identify logs coming from the paper trading internal logic easily
- the way the fees are calculated, the margin fees will be used for all orders on the short exchange, even if margin is not needed (if we have sufficient funds to sell, or buy I think we should only pay the trade fees)

I don't now if it's how it's supposed to work: if we set the leverage when buying on the short exchange, are we supposed to pay margin fees? I would think so, but then the setLeverage(2) in the TradingService.executeOrderPair method should not be set for the short exchange BUY orders - as we don't need the leverage.

A better way to do this would be to check if a margin trade is required or not 
Create 
```
boolean isMarginRequired (LimitOrder order) {
        BigDecimal counterDelta = getCounterDelta(order);
        BigDecimal baseDelta = getBaseDelta(order);
 
        //Margin required because we don't have enough cash
        if(exchange.getPaperAccountService().getBalance(order.getCurrencyPair().counter).add(counterDelta).compareTo(BigDecimal.ZERO) <0)
            return true;

        //Margin required because we dont have enough crypto 
        if(exchange.getPaperAccountService().getBalance(order.getCurrencyPair().base).add(baseDelta).compareTo(BigDecimal.ZERO) <0)
            return true;

        //Margin required because the leverage was set
        if(order.getLeverage() != null)
            return true;

        return false;
    }
```
Change checkBalance (LimitOrder order) to
```
private void checkBalance(LimitOrder order) {
        if(isMarginRequired(order) && !exchangeService.getExchangeMetadata(exchange).getMargin())
            throw new FundsExceededException();
    }
```
And use `isMarginRequired` to see if we should use the margin fees.
```
fee = isMarginRequired(order) && exchangeFee.getMarginFee().isPresent() ? exchangeFee.getMarginFee().get() : exchangeFee.getTradeFee();
LOGGER.info("{} paper exchange: margin enabled:{}|margin required:{}| order using {} fee", exchangeName, exchangeConfiguration.getMargin(), isMarginRequired(order), fee);
```

Reply:
Thanks for the nice feedback :+1: 

I think we always pay margin fees even if we have enough balance for the borrowed coins (short), @scionaltera can you please confirm?

As far as I know setting value`2` to `order.getLeverage()` is only needed for Liquid exchange. @scionaltera can you confirm this as well?

Reply:
The exchange should decide whether it's a margin or non-margin trade by looking at `LimitOrder.getLeverage()`. If it's not null, then it's a margin trade (regardless of long or short) and margin fees should be charged. If we really wanted to do it right for margin orders we wouldn't take the full amount out of the balance either. On a 2:1 trade (`getLeverage() == 2`) we'd only take half of the value away from the balance. It's not that important for Arbitrader at this time because we only ever set it to 2 and we never trade above our actual balance.

Reply:
In that case, we can have a method `useMargin` to determine if we charge the margin fees or not.

```
private void useMargin(LimitOrder order) {
    return order.getLeverage() != null && exchangeService.getExchangeMetadata(exchange).getMargin() && exchangeFee.getMarginFee().isPresent() ;
}
```
The fees could be:
```
fee = useMargin(order) ? exchangeFee.getMarginFee().get() : exchangeFee.getTradeFee();
LOGGER.info("{} paper exchange: margin enabled:{}|margin required:{}| order using {} fee", exchangeName, exchangeConfiguration.getMargin(), useMargin(order), fee);
```

and change the checkBalance method accordingly

```
private void checkBalance(LimitOrder order) {
    if(!useMargin(order)) {
        BigDecimal counterDelta = getCounterDelta(order);
        BigDecimal baseDelta = getBaseDelta(order);
 
        //Margin required because we don't have enough cash
        if(exchange.getPaperAccountService().getBalance(order.getCurrencyPair().counter).add(counterDelta).compareTo(BigDecimal.ZERO) <0)
            throw new FundsExceededException();

        //Margin required because we dont have enough crypto 
        if(exchange.getPaperAccountService().getBalance(order.getCurrencyPair().base).add(baseDelta).compareTo(BigDecimal.ZERO) <0)
            throw new FundsExceededException();
    } else {
         //TODO after implementing leverage in the paper exchange, we should check here if we have sufficient funds to cover the leverage
    }
}
```

Reply:
One other wrinkle I just thought of - the PaperExchange should probably check whether it's configured with `margin: true` or not and reject margin orders if it is not supposed to be a margin-capable exchange. It would be a bug if we tried to submit a margin order to a non-margin exchange, but stranger things have happened.

Reply:
It does check it with exchangeService.getExchangeMetadata(exchange).getMargin(). Currently, it rejects orders that would make any balance negative unless margin is enabled, via the check balance method

Reply:
Thanks @AllirionX for your great feedback!
@scionaltera yeah, I have implemented a check condition for that use case.